File: examples/MDP/tiger_antelope.cpp
/* This file contains the code that is presented in the Tutorial.md
 * file in the main folder of this project. The tutorial can also
 * be viewed in the main page of the doxygen documentation.
 *
 * This code implements a problem where a tiger needs to plan
 * in order to catch an antelope in a NxN toroidal grid world.
 *
 * The implementation is not efficient since all transition
 * probabilities are computed on the fly; storing them in a
 * matrix would make the solver work faster. Modifying the code
 * to allow this is trivial, and I wanted to keep the example
 * simple in order to introduce the theory behind the library.
 *
 * For more examples be sure to check out the "tests" folder!
 * The code there is very simple and it contains most usages
 * of this library ever, and it will probably give you an
 * even better introduction than this code does.
 */
#include <iostream>
#include <iomanip>
#include <fstream>
#include <array>
#include <cmath>
#include <chrono>
#include <thread>

#include <AIToolbox/MDP/Algorithms/ValueIteration.hpp>
#include <AIToolbox/MDP/Policies/Policy.hpp>
#include <AIToolbox/MDP/IO.hpp>
#include <AIToolbox/MDP/SparseModel.hpp>

// MODEL

constexpr int SQUARE_SIZE = 8;

using CoordType = std::array<int, 4>;
enum {
    TIGER_X = 0,
    TIGER_Y = 1,
    ANTEL_X = 2,
    ANTEL_Y = 3
};

// Returns distance between coordinates. It is consistent with
// the wraparound world.
int wrapDiff( int coord1, int coord2 ) {
    int diff = coord2 - coord1;

    int distance1 = std::abs( diff ), distance2 = SQUARE_SIZE - distance1;
    if ( distance1 < distance2 ) return diff;
    return diff > 0 ? -distance2 : distance2;
}

size_t A = 5;
enum {
    UP    = 0,
    DOWN  = 1,
    LEFT  = 2,
    RIGHT = 3,
    STAND = 4
};

double getTransitionProbability( const CoordType & c1, size_t action, const CoordType & c2 ) {
    // We compute the distances traveled by both the antelope and the tiger.
    int tigerMovementX = wrapDiff( c1[TIGER_X], c2[TIGER_X] );
    int tigerMovementY = wrapDiff( c1[TIGER_Y], c2[TIGER_Y] );
    int antelMovementX = wrapDiff( c1[ANTEL_X], c2[ANTEL_X] );
    int antelMovementY = wrapDiff( c1[ANTEL_Y], c2[ANTEL_Y] );

    // Both the tiger and the antelope can only move by 1 cell max at each
    // timestep. Thus, if this is not the case, the transition is
    // impossible.
    if ( std::abs( tigerMovementX ) +
         std::abs( tigerMovementY ) > 1 ) return 0.0;

    if ( std::abs( antelMovementX ) +
         std::abs( antelMovementY ) > 1 ) return 0.0;

    // Now we check whether the tiger was next to the antelope or not
    int diffX = wrapDiff( c1[TIGER_X], c1[ANTEL_X] );
    int diffY = wrapDiff( c1[TIGER_Y], c1[ANTEL_Y] );

    // We check whether they were both in the same cell before.
    // In that case the game would have ended, and nothing would happen anymore.
    // We model this as a self-absorbing state, or a state that always transitions
    // to itself. This is valid no matter the action taken.
    if ( diffX == 0 && diffY == 0 ) {
        if ( c1 == c2 ) return 1.0;
        else return 0.0;
    }

    // The tiger can move only in the direction specified by its action. If
    // it is not the case, the transition is impossible.
    if ( action == STAND && ( tigerMovementX || tigerMovementY ) ) return 0.0;
    if ( action == UP    && tigerMovementY != 1  ) return 0.0;
    if ( action == DOWN  && tigerMovementY != -1 ) return 0.0;
    if ( action == LEFT  && tigerMovementX != -1 ) return 0.0;
    if ( action == RIGHT && tigerMovementX != 1  ) return 0.0;

    // If they were not adjacent, then the probability for any move of the
    // antelope is simply 1/5: it behaves randomly.
    if ( std::abs( diffX ) + std::abs( diffY ) > 1 ) return 1.0 / 5.0;

    // Otherwise, first we check that the move was allowed, as
    // the antelope cannot move where the tiger was before.
    if ( c1[TIGER_X] == c2[ANTEL_X] && c1[TIGER_Y] == c2[ANTEL_Y] ) return 0.0;

    // Else the probability of this transition is 1 / 4, still random but without
    // a possible antelope action.
    return 1.0 / 4.0;
}

double getReward( const CoordType & c ) {
    if ( c[TIGER_X] == c[ANTEL_X] && c[TIGER_Y] == c[ANTEL_Y] ) return 10.0;
    return 0.0;
}

constexpr double discount = 0.9;

size_t encodeState(const CoordType & coords) {
    size_t state = 0; unsigned multiplier = 1;
    for ( auto c : coords ) {
        state += multiplier * c;
        multiplier *= SQUARE_SIZE;
    }
    return state;
}

CoordType decodeState(size_t state) {
    CoordType coords;
    for ( auto & c : coords ) {
        c = state % SQUARE_SIZE;
        state /= SQUARE_SIZE;
    }
    return coords;
}

class GridWorld {
    public:
        size_t getS() const { return SQUARE_SIZE * SQUARE_SIZE * SQUARE_SIZE * SQUARE_SIZE; }
        size_t getA() const { return ::A; }
        double getDiscount() const { return ::discount; }

        double getTransitionProbability( size_t s, size_t a, size_t s1 ) const {
            return ::getTransitionProbability( decodeState( s ), a, decodeState( s1 ) );
        }

        double getExpectedReward( size_t, size_t, size_t s1 ) const {
            return ::getReward( decodeState( s1 ) );
        }

        // These two functions are needed to keep template code in the library
        // simple, but you don't need to implement them for the method we use
        // in this example.
        std::tuple<size_t, double> sampleSR(size_t,size_t) const;
        bool isTerminal(size_t) const;
};

// RENDERING

// Special character to go back up when drawing.
std::string up =   "\033[XA";
// Special character to go back to the beginning of the line.
std::string back = "\33[2K\r";

void goup(unsigned x) {
    while (x > 9) {
        up[2] = '0' + 9;
        std::cout << up;
        x -= 9;
    }
    up[2] = '0' + x;
    std::cout << up;
}

void godown(unsigned x) {
    while (x) {
        std::cout << '\n';
        --x;
    }
}

void printState(const CoordType & c) {
    for ( int y = SQUARE_SIZE - 1; y >= 0; --y ) {
        for ( int x = 0; x < SQUARE_SIZE; ++x ) {
            if (x == c[TIGER_X] && y == c[TIGER_Y]) std::cout << "@ ";
            else if (x == c[ANTEL_X] && y == c[ANTEL_Y]) std::cout << "A ";
            else std::cout << ". ";
        }
        std::cout << std::endl;
    }
}

void printCurrentTimeString() {
    auto t = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    auto tm = *std::localtime(&t);
    std::cout << std::put_time(&tm, "%T");
}

int main() {
    GridWorld world;

    // This is optional, and should make solving the model almost instantaneous.
    // Unfortunately, since our main model is so big, the copying process
    // still takes a lot of time. But at least that would be a one-time cost!
    printCurrentTimeString(); std::cout << " - Constructing MDP...\n";
    AIToolbox::MDP::SparseModel model(world);

    // This is a method that solves MDPs completely. It has a couple of
    // parameters available.
    // The only non-optional parameter is the horizon of the solution; as in
    // how many steps should the solution look ahead in order to decide which
    // move to take. If we chose 1, for example, the tiger would only consider
    // cells next to it to decide where to move; this wouldn't probably be
    // what we want.
    // We want the tiger to think for infinite steps: this can be
    // approximated with a very high horizon, since in theory the final solution
    // will converge to a single policy anyway. Thus we put a very high number
    // as the horizon here.
    printCurrentTimeString(); std::cout << " - Solving MDP using infinite horizon...\n";
    AIToolbox::MDP::ValueIteration solver(1000000);

    // This is where the magic happen. This could take around 10-20 minutes,
    // depending on your machine (most of the time is spent on this tutorial's
    // code, however, since it is a pretty inefficient implementation).
    // But you can play with it and make it better!
    //
    // If you are using the Sparse Model though, it is instantaneous since
    // Eigen is very very efficient in computing the values we need!
    auto solution = solver(model);

    printCurrentTimeString();
    std::cout << " - Converged: " << (std::get<0>(solution) < solver.getTolerance()) << "\n";

    AIToolbox::MDP::Policy policy(world.getS(), world.getA(), std::get<1>(solution));

    // We create a random engine to pick a random state as start.
    std::default_random_engine rand(AIToolbox::Impl::Seeder::getSeed());
    std::uniform_int_distribution<size_t> start(0, SQUARE_SIZE * SQUARE_SIZE * SQUARE_SIZE * SQUARE_SIZE - 1);

    size_t s, a, s1;
    double r, totalReward = 0.0;

    // We create a starting state which is not the end.
    do s = start(rand);
    while (model.isTerminal(s));

    size_t t = 100;
    while (true) {
        // Print it!
        printState(decodeState(s));

        // We give the tiger a time limit, but if it reaches
        // the antelope we end the game.
        if (t == 0 || model.isTerminal(s)) break;

        // We sample an action for this state according to the optimal policy
        a = policy.sampleAction(s);
        // And we use the model to simulate what is going to happen next (in
        // case of a "real world" scenario where the library is used this step
        // would not exist as the world would automatically step to the next
        // state. Here we simulate.
        std::tie(s1, r) = model.sampleSR(s, a);

        // Add into the total reward (we don't use this here, it's just as an
        // example)
        totalReward += r;
        // Update the current state with the new one.
        s = s1;

        --t;
        goup(SQUARE_SIZE);

        // Sleep 1 second so the user can see what is happening.
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    // You can save, and then load up this policy again on files. You will not
    // need to solve the model again ever, and you can embed the policy into
    // any application you want!
    // {
    //     std::ofstream output("policy.txt");
    //     output << policy;
    // }

    return 0;
}

File: examples/MDP/tiger_antelope.py
# This example shows how to perform Value Iteration for a simple single
# agent game, modeled as MDP. It computes the value and Q-value, and shows for
# the optimal policy for sample states.
#
# Suppose you have a grid shaped world, of N x N cells by. The world loops
# on itself like a torus, so that the top and bottom borders are connected
# with each other, and the same is true for the left and right borders.
# Two creatures walk this world: a tiger, and an antelope. Both creatures
# can move in the following way: up, down, left or right or stand still.
# When they decide to move, their movement is deterministic. The two
# creatures have different goals.
#
# The goal of the antelope is to not get eaten by the tiger. However it is
# pretty clueless in doing so; in fact, it always moves or stands randomly,
# aside from when the tiger is directly next to it. In that case, it will
# move randomly anywhere, but towards the tiger.
#
# The tiger has the goal of catching the antelope. Once it catches it,
# the game ends. What would be the best way for it to move?

import argparse
import itertools
import os
import sys
import time
from random import randint

build_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                         '..', 'build')
if build_dir not in sys.path:
    sys.path.append(build_dir)

try:
    from AIToolbox import MDP
except ImportError:
    raise ImportError("cannot find MDP, has it been built?")

# Model

def wrapDiff(coord1, coord2):
    """
    Compute difference between two coordinates, consistent with the
    wraparound world with size SQUARE_SIZE x SQUARE_SIZE.

    Parameters
    ----------
    coord1: int
        First coordinate.
    coord2: int
        Second coordinate.
    """
    diff = coord2 - coord1
    distance1 = abs(diff)
    distance2 = SQUARE_SIZE - distance1

    if distance1 < distance2:
        return diff
    elif diff > 0:
        return -distance2
    return distance2


def getTransitionProbability(coord1, action, coord2):
    """
    Given two coordinates and an action, return the transition probability.

    Parameters
    ----------
    coord1: tuple of int
        Four element tuple containing the current coordinates of the tiger and
        antelope.
    action: str
        String representing the tiger action, options are i
        'stand', 'up', 'down', 'left', 'right'.
    coord2: tuple of int
        Four element tuple containing the next coordinates of the tiger and
        antelope.
    """
    # We compute the distances traveled by both the antelope and the tiger.
    tiger_x1, tiger_y1, antelope_x1, antelope_y1 = coord1
    tiger_x2, tiger_y2, antelope_x2, antelope_y2 = coord2

    # We compute the distances traveled by both the antelope and the tiger.
    tigerMovementX = wrapDiff(tiger_x1, tiger_x2)
    tigerMovementY = wrapDiff(tiger_y1, tiger_y2)
    antelMovementX = wrapDiff(antelope_x1, antelope_x2)
    antelMovementY = wrapDiff(antelope_y1, antelope_y2)

    # Both the tiger and the antelope can only move by 1 cell max at each
    # timestep. Thus, if this is not the case, the transition is
    # impossible.
    if abs(tigerMovementX) + abs(tigerMovementY) > 1:
        return 0.0
    if abs(antelMovementX) + abs(antelMovementY) > 1:
        return 0.0

    # Now we check whether the tiger was next to the antelope or not
    diffX = wrapDiff(tiger_x1, antelope_x1)
    diffY = wrapDiff(tiger_y1, antelope_y1)

    # We check whether they were both in the same cell before.
    # In that case the game would have ended, and nothing would happen anymore.
    # We model this as a self-absorbing state, or a state that always
    # transitions to itself. This is valid no matter the action taken.
    if diffX == 0 and diffY == 0:
        if coord1 == coord2:
            return 1.0
        return 0.0

    # The tiger can move only in the direction specified by its action. If
    # it is not the case, the transition is impossible.
    if action == 'stand' and (tigerMovementX or tigerMovementY):
        return 0.0
    if action == 'up' and tigerMovementY != 1:
        return 0.0
    if action == 'down' and tigerMovementY != -1:
        return 0.0
    if action == 'left' and tigerMovementX != -1:
        return 0.0
    if action == 'right' and tigerMovementX != 1:
        return 0.0

    # If tiger and antelope are not adjacent, then the probability for any
    # move of the antelope is simply 1/5: it behaves randomly.
    if abs(diffX) + abs(diffY) > 1:
        return 1.0 / 5.0

    # Otherwise, first we check that the move was allowed, as
    # the antelope cannot move where the tiger was before.
    if tiger_x1 == antelope_x2 and tiger_y1 == antelope_y2:
        return 0.0

    # Else the probability of this transition is 1 / 4, still random but
    # without a possible antelope action.
    return 1.0 / 4.0


def getReward(coord):
    """
    Given a state, return the reward.

    Parameters
    ----------
    coord: tuple of int
        Four element tuple containing the position of the tiger and antelope.

    Returns
    -------
    reward: float
        If the tiger catches the antelope, reward is +10. If not, reward is 0.0
    """
    tiger_x, tiger_y, antelope_x, antelope_y = coord
    if tiger_x == antelope_x and tiger_y == antelope_y:
        return 10.0
    return 0.0


def encodeState(coord):
    """
    Convert from coordinate to state_index.

    Parameters
    ----------
    coord: tuple of int
        Four element tuple containing the position of the tiger and antelope.

    Returns
    -------
    state: int
        Index of the state.
    """
    state = 0
    multiplier = 1
    for c in coord:
        state += multiplier * c
        multiplier *= SQUARE_SIZE

    return state


def decodeState(state):
    """
    Convert from state_index to coordinate.

    Parameters
    ----------
    state: int
        Index of the state.

    Returns
    -------
    coord: tuple of int
        Four element tuple containing the position of the tiger and antelope.
    """
    coord = []
    for _ in range(4):
        c = state % SQUARE_SIZE
        state /= SQUARE_SIZE
        coord.append(c)
    return tuple(coord)


#RENDERING

# Special character to go back up when drawing.
up = list("\033[XA")
# Special character to go back to the beginning of the line.
back = list("\33[2K\r")

def goup(x):
    """ Moves the cursor up by x lines """
    while x > 8:
        up[2] = '9'
        print "".join(up)
        x -= 8

    up[2] = str(x + 1)
    print "".join(up)

def godown(x):
    """ Moves the cursor down by x lines """
    while x:
        print ""
        x -= 1

def printState(coord):
    """
    Draw the grid world.

    - @ represents the tiger.
    - A represents the antelope.

    Parameters
    ----------
    coord: tuple of int
        Four element tuple containing the position of the tiger and antelope.
    """
    t_x, t_y, a_x, a_y = coord
    for y in range(SQUARE_SIZE-1, -1, -1):
        for x in range(SQUARE_SIZE):
            if (t_x, t_y) == (x, y):
                print "@",
            elif (a_x, a_y) == (x, y):
                print "A",
            else:
                print ".",
        print ""

def solve_mdp(horizon, tolerance, discount=0.9):
    """
    Construct the gridworld MDP, and solve it using value iteration. Print the
    best found policy for sample states.

    Returns
    -------
    solution: tuple
        First element is a boolean that indicates whether the method has
        converged. The second element is the value function. The third
        element is the Q-value function, from which a policy can be derived.
    """
    print time.strftime("%H:%M:%S"), "- Constructing MDP..."

    # Statespace contains the tiger (x, y) and antelope (x, y). Note that
    # this is a very naive state representation: many of these states can be
    # aggregated! We leave this as an exercise to the reader :)
    # S = [(t_x, t_y, a_x, a_y), .. ]
    S = list(itertools.product(range(SQUARE_SIZE), repeat=4))

    # A = tiger actions
    A = ['stand', 'up', 'down', 'left', 'right']

    # T gives the transition probability for every s, a, s' triple.
    T = []
    for state in range(len(S)):
        coord = decodeState(state)
        T.append([[getTransitionProbability(coord, action,
                                            decodeState(next_state))
                   for next_state in range(len(S))] for action in A])

    # R gives the reward associated with every s, a, s' triple. In the current
    # example, we only specify reward for s', but we still need to give the
    # entire |S|x|A|x|S| array.
    reward_row = [getReward(decodeState(next_state))
                  for next_state in range(len(S))]
    R = [[reward_row for _ in A] for _ in S]

    # set up the model
    model = MDP.SparseModel(len(S), len(A))
    model.setTransitionFunction(T)
    model.setRewardFunction(R)
    model.setDiscount(discount)

    # Perform value iteration
    print time.strftime("%H:%M:%S"), "- Solving MDP using ValueIteration(horizon={}, tolerance={})".format(
        horizon, tolerance)

    solver = MDP.ValueIteration(horizon, tolerance)
    solution = solver(model)

    print time.strftime("%H:%M:%S"), "- Converged:", solution[0] < solver.getTolerance()
    _, value_function, q_function = solution

    policy = MDP.Policy(len(S), len(A), value_function)

    s = randint(0, SQUARE_SIZE**4 - 1)
    while model.isTerminal(s):
        s = randint(0, SQUARE_SIZE**4 - 1)

    totalReward = 0
    for t in xrange(100):
        printState(decodeState(s))

        if model.isTerminal(s):
            break

        a = policy.sampleAction(s)
        s1, r = model.sampleSR(s, a)

        totalReward += r
        s = s1

        goup(SQUARE_SIZE)

        state = encodeState(coord)

        # Sleep 1 second so the user can see what is happening.
        time.sleep(1)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('-s', '--square-size', default=5, type=int,
                        help="Size of the square gridworld.")
    parser.add_argument('-ho', '--horizon', default=1000000, type=int,
                        help="Horizon parameter for value iteration")
    parser.add_argument('-t', '--tolerance', default=0.001, type=float,
                        help="Tolerance parameter for value iteration")
    parser.add_argument('-d', '--discount', default=0.9, type=float,
                        help="Discount parameter for value iteration")

    args = parser.parse_args()
    SQUARE_SIZE = args.square_size
    solve_mdp(horizon=args.horizon, tolerance=args.tolerance)
